aesd-char-driver/main.c

```diff
@@ -63,71 +66,139 @@ ssize_t aesd_read(struct file *filp, char __user *buf, size_t count, loff_t *f_p
-    static size_t final_count = 0;
+
+    struct aesd_dev *char_dev = filp->private_data;
 
     /*Temporary buffer to store the string from user*/
     char* temp_buffptr = kmalloc(count, GFP_KERNEL);
     copy_from_user(temp_buffptr, buf, count);
 
     /*If newline character not found, append to final_buffptr*/
-    if(strchr(temp_buffptr, '\n') == NULL) {
-        final_buffptr = krealloc(final_buffptr, final_count + count, GFP_KERNEL); //realloc based on new size
-        memcpy(final_buffptr + final_count, temp_buffptr, count);
-        PDEBUG("final_buffptr: %s", final_buffptr);
-        final_count += count;
-        retval = count;
+    if(memchr(temp_buffptr, '\n', count) == NULL) {
+        if(char_dev->final_buffptr == NULL) {
+            char_dev->final_buffptr = kmalloc(count, GFP_KERNEL);
+        }
+        else {
+            char_dev->final_buffptr = krealloc(char_dev->final_buffptr, char_dev->final_count + count, GFP_KERNEL);
+        }
+        memcpy(char_dev->final_buffptr + char_dev->final_count, temp_buffptr, count);
+        PDEBUG("final_buffptr: %s", char_dev->final_buffptr);
+        char_dev->final_count += count;
         kfree(temp_buffptr); //this is not needed anymore
         temp_buffptr = NULL; // Dont leave it dangling
-        return retval;
     }
 
-    /*If newline character found, append to final_buffptr and write to buffer*/
-    final_buffptr = krealloc(final_buffptr, final_count + count, GFP_KERNEL);
-    memcpy(final_buffptr + final_count, temp_buffptr, count);
-    PDEBUG("final_buffptr: %s", final_buffptr);
-    final_count += count;
-
-    /*Write to the circular command buffer*/
-    struct aesd_buffer_entry *new_entry = kmalloc(sizeof(struct aesd_buffer_entry), GFP_KERNEL);
-    if (new_entry == NULL) {
-        retval = -ENOMEM;
-    }
     else {
-        new_entry->size = final_count;
-        new_entry->buffptr = kmalloc(final_count, GFP_KERNEL);
-        memcpy(new_entry->buffptr, final_buffptr, final_count);
+        /*If newline character found, append to final_buffptr and write to buffer*/
+        if(char_dev->final_buffptr == NULL) {
+            char_dev->final_buffptr = kmalloc(count, GFP_KERNEL);
+        }
+        else {
+            char_dev->final_buffptr = krealloc(char_dev->final_buffptr, char_dev->final_count + count, GFP_KERNEL);
+        }
+        memcpy(char_dev->final_buffptr + char_dev->final_count, temp_buffptr, count);
+        PDEBUG("final_buffptr: %s", char_dev->final_buffptr);
+        char_dev->final_count += count;
+
+        /*Write to the circular command buffer*/
+        struct aesd_buffer_entry new_entry;
+        new_entry.size = char_dev->final_count;
+        new_entry.buffptr = kmalloc(char_dev->final_count, GFP_KERNEL);
+        memcpy(new_entry.buffptr, char_dev->final_buffptr, char_dev->final_count);
 
         /*Lock before writing*/
-        mutex_lock(&aesd_device.write_lock);
+        mutex_lock(&char_dev->read_write_lock);
```

Comment:
@ashwinhr186 you need locking for `final_buffptr` - this lock needs to be acquired before line 84.